### PR TITLE
Lazy

### PR DESCRIPTION
**Pull request by JohannesRabauer**

### CODE REVIEW

File: docs/modules/ROOT/pages/lazies.adoc [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+	//...
+----
+
+== Internas
+
+SpringDataEclipseStoreLazies work as a proxy for the EclipseStore-Lazies.
+As far as EclipseStore is concerned, a SpringDataEclipseStoreLazy-Object is a normal Java object that contains a Lazy-Reference.
+
+But when {product-name} creates the working copy, *the SpringDataEclipseStoreLazy-Reference is not resolved* but instead only a reference to the original Lazy-Object in EclipseStore is loaded.
+As soon as ``get()`` is called on the SpringDataEclipseStoreLazy, a *new working copy of the lazy object* is created.
```
Comment (tboehm-xdev) [inline docs/modules/ROOT/pages/lazies.adoc:40]: Maybe add a comment what happens if the original `Lazy` class of EclipseStore is used.

File: spring-data-eclipse-store/src/main/java/software/xdev/spring/data/eclipse/store/exceptions/LazyNotUnlinkableException.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package software.xdev.spring.data.eclipse.store.exceptions;
+
+public class LazyNotUnlinkableException extends RuntimeException
```
Comment (tboehm-xdev) [inline spring-data-eclipse-store/src/main/java/software/xdev/spring/data/eclipse/store/exceptions/LazyNotUnlinkableException.java:18]: Maybe add some documentation when this exception should be raised or when not

File: spring-data-eclipse-store/src/main/java/software/xdev/spring/data/eclipse/store/repository/lazy/SpringDataEclipseStoreLazy.java [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+			{
+				this.wrappedLazy = this.createNewDefaultLazyWithClearableReference();
+			}
+			return this.wrappedLazy;
+		}
+		
+		@SuppressWarnings("unchecked")
+		private Lazy<T> createNewDefaultLazyWithClearableReference()
+		{
+			Objects.requireNonNull(this.loader);
```
Comment (tboehm-xdev) [inline spring-data-eclipse-store/src/main/java/software/xdev/spring/data/eclipse/store/repository/lazy/SpringDataEclipseStoreLazy.java:108]: Maybe move those `requireNonNull` lines to constructor?

File: spring-data-eclipse-store/src/main/java/software/xdev/spring/data/eclipse/store/repository/lazy/SpringDataEclipseStoreLazy.java [hunk 4]
```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright © 2023 XDEV Software (https://xdev.software)
```
Comment (tboehm-xdev) [inline spring-data-eclipse-store/src/main/java/software/xdev/spring/data/eclipse/store/repository/lazy/SpringDataEclipseStoreLazy.java:2]: Maybe update to 2024 (also in other files)

File: spring-data-eclipse-store/src/main/java/software/xdev/spring/data/eclipse/store/repository/lazy/SpringDataEclipseStoreLazyBinaryHandler.java [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+ *     First case:<br/>
+ *     The user creates a {@link SpringDataEclipseStoreLazy} and puts a object in it.
+ *     This object is stored as with a default {@link BinaryTypeHandler}. But when it gets loaded,
+ *     it <b>does not</b> load as the stored object, but it gets wrapped in a {@link Lazy#Reference(Object)}.
+ * </p>
+ * <p>
+ *     Second case:<br/>
+ *     The actual lazy object gets loaded from the actual storage. In this case the {@link ObjectSwizzling} is
+ *     important! It's the actual {@link ObjectSwizzling} from the storage (not from the
+ * {@link software.xdev.spring.data.eclipse.store.repository.support.copier.copier.EclipseSerializerRegisteringCopier}).
```
Comment (tboehm-xdev) [inline spring-data-eclipse-store/src/main/java/software/xdev/spring/data/eclipse/store/repository/lazy/SpringDataEclipseStoreLazyBinaryHandler.java:48]: Package contains two times `copier`?

File: spring-data-eclipse-store/src/main/java/software/xdev/spring/data/eclipse/store/repository/lazy/SpringDataEclipseStoreLazyBinaryHandler.java [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+	static final Constructor<SpringDataEclipseStoreLazy.Default> CONSTRUCTOR_SURROGATE_LAZY = XReflect.setAccessible(
+		XReflect.getDeclaredConstructor(
+			SpringDataEclipseStoreLazy.Default.class,
+			long.class,
+			ObjectSwizzling.class,
+			WorkingCopier.class
+		)
+	);
+	
+	public static final int OFFSET_UNWRAPPED_OBJECT = 8;
```
Comment (tboehm-xdev) [inline spring-data-eclipse-store/src/main/java/software/xdev/spring/data/eclipse/store/repository/lazy/SpringDataEclipseStoreLazyBinaryHandler.java:68]: `public` visibility necessary?

File: spring-data-eclipse-store/src/main/java/software/xdev/spring/data/eclipse/store/repository/lazy/SpringDataEclipseStoreLazyBinaryHandler.java [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+	
+	private final ObjectSwizzling originalStoreLoader;
+	private final WorkingCopier<?> copier;
+	
+	public SpringDataEclipseStoreLazyBinaryHandler(
+		final ObjectSwizzling originalStoreLoader,
+		final WorkingCopier<?> copier)
+	{
+		super(
+			(Class)SpringDataEclipseStoreLazy.Default.class,
```
Comment (tboehm-xdev) [inline spring-data-eclipse-store/src/main/java/software/xdev/spring/data/eclipse/store/repository/lazy/SpringDataEclipseStoreLazyBinaryHandler.java:79]: Cast necessary? If yes, maybe add some docs

File: spring-data-eclipse-store/src/main/java/software/xdev/spring/data/eclipse/store/repository/lazy/SpringDataEclipseStoreLazyBinaryHandler.java [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+			CustomFields(
+				CustomField(Object.class, "lazySubject"),
+				CustomField(Object.class, "unwrappedSubject")
+			)
+		);
+		this.originalStoreLoader = Objects.requireNonNull(originalStoreLoader);
+		this.copier = copier;
```
Comment (tboehm-xdev) [inline spring-data-eclipse-store/src/main/java/software/xdev/spring/data/eclipse/store/repository/lazy/SpringDataEclipseStoreLazyBinaryHandler.java:86]: No nullcheck required?

File: spring-data-eclipse-store/src/main/java/software/xdev/spring/data/eclipse/store/repository/lazy/SpringDataEclipseStoreLazyBinaryHandler.java (hunk 6 above)
Comment (tboehm-xdev) [reply spring-data-eclipse-store/src/main/java/software/xdev/spring/data/eclipse/store/repository/lazy/SpringDataEclipseStoreLazyBinaryHandler.java:68]: Are those constants maybe available from eclipsestore?

File: spring-data-eclipse-store/src/main/java/software/xdev/spring/data/eclipse/store/repository/lazy/SpringDataEclipseStoreLazy.java (hunk 3 above)
Comment (JohannesRabauer) [reply spring-data-eclipse-store/src/main/java/software/xdev/spring/data/eclipse/store/repository/lazy/SpringDataEclipseStoreLazy.java:108]: This is not valid. In the constructor it is valid to leave the loader to null (hence the other constructors). But the method is not allowed to get executed when the loader is null.

File: spring-data-eclipse-store/src/main/java/software/xdev/spring/data/eclipse/store/repository/lazy/SpringDataEclipseStoreLazyBinaryHandler.java (hunk 6 above)
Comment (JohannesRabauer) [reply spring-data-eclipse-store/src/main/java/software/xdev/spring/data/eclipse/store/repository/lazy/SpringDataEclipseStoreLazyBinaryHandler.java:68]: I think they are available, but should not be used. This is a pattern, that is common in the binary handler and should only be used locally. The value could change in the future and then should also only affect this class.

**Pull request merged**